test/E2ETest/Microsoft.Test.E2E.AspNet.OData/BulkOperation/BulkOperationPatchHandlers.cs

```diff
@@ -9,107 +9,148 @@
 using System.Collections.Generic;
 using System.Linq;
 using Microsoft.AspNet.OData;
+using Microsoft.AspNet.OData.Common;
+using Microsoft.AspNet.OData.Extensions;
 using Microsoft.OData.Edm;
-using Microsoft.Test.E2E.AspNet.OData.BulkInsert;
+using Microsoft.OData.UriParser;
 
 namespace Microsoft.Test.E2E.AspNet.OData.BulkOperation
 {
     internal class APIHandlerFactory : ODataAPIHandlerFactory
     {
-        public APIHandlerFactory(IEdmModel model) : base (model)
+        public APIHandlerFactory(IEdmModel model) : base(model)
         {
         }
 
-        public override IODataAPIHandler GetHandler(NavigationPath navigationPath)
+        public override IODataAPIHandler GetHandler(ODataPath odataPath)
         {
-            if(navigationPath != null)
+            if (odataPath != null)
             {
-                var pathItems = navigationPath;
+                int currentPosition = 0;
 
-                int cnt = 0;
-                                    
-                    switch (pathItems[cnt].Name)
-                    {
-                        case "Employees":
-                        {
-                            Employee employee;
-                            string msg;
-                            if ((new EmployeeAPIHandler().TryGet(pathItems[cnt].KeyProperties, out employee, out msg)) == ODataAPIResponseStatus.Success)
-                            {
-                                return GetNestedHandlerForEmployee(pathItems, cnt, employee);
-                            }
-                        }
-                        return null;
-                    case "Companies":
-                        return new CompanyAPIHandler();
+                if (odataPath.Count == 1)
+                {
+                    GetHandlerInternal(odataPath.FirstSegment.Identifier, currentPosition);
+                }
 
-                        default:
-                            return null;
+                List<ODataPathSegment> pathSegments = odataPath.GetSegments();
 
-                    }
-                
+                ODataPathSegment currentPathSegment = pathSegments[currentPosition];
+
+                if (currentPathSegment is EntitySetSegment || currentPathSegment is NavigationPropertySegment || currentPathSegment is SingletonSegment)
+                {
+                    int keySegmentPosition = ODataPathHelper.GetNextKeySegmentPosition(pathSegments, currentPosition);
+                    KeySegment keySegment = (KeySegment)pathSegments[keySegmentPosition];
```

Comment:
will the user be writing this logic? Can this be moved to the `ODataHelper` class?

Reply:
If i understand this correctly, here we are trying to get the key position for the current path segment right? why do we need to know the keysegment position?

Reply:
I think i have a little bit of understanding here..so if we have ''Companies(1)/Employees(2).
so we'll have EntitySetSegment/KeySegment/EntitySetSegment/KeySegment. 

when we find the first `EntitySetSegment`, we get it's key segment and it's position.. so when we'll be getting the nestedhandler,, we increment the keyposition by one and use that as the currentposition to get the nestedhandler for the nested item.

Reply:
so what if we had.. Companies(1)/typesegment or containedsegment - will this logic still apply? I understand this is user implementation but we can also have examples to show these other cases.